keeper/abci.go

```diff
@@ -22,12 +22,172 @@ package keeper
+		}
+
+		if !navInfo.LastUpdate.IsZero() {
+			state.LastNavUpdate = navInfo.LastUpdate
+		}
+
+		return k.SetVaultsV2VaultState(ctx, state)
+	}
+
+	if state.TotalNav.Equal(navInfo.CurrentNav) && !navInfo.LastUpdate.After(state.LastNavUpdate) {
+		return nil
+	}
+
+	navBig := navInfo.CurrentNav.BigInt()
+	totalSharesBig := totalShares.BigInt()
+
+	residual := new(big.Int)
+	aggregatedYield := sdkmath.ZeroInt()
+	aggregatedDeposits := sdkmath.ZeroInt()
+
+	err = k.IterateVaultsV2UserPositions(ctx, func(address types.AccAddress, position vaultsv2.UserPosition) (bool, error) {
+		shares, err := k.GetVaultsV2UserShares(ctx, address)
+		if err != nil {
+			return true, err
+		}
+
+		if !shares.IsPositive() {
+			pending := position.AmountPendingWithdrawal
+
+			if !position.AccruedYield.IsZero() || position.DepositAmount.IsNil() || !position.DepositAmount.Equal(pending) {
+				position.AccruedYield = sdkmath.ZeroInt()
+				position.DepositAmount = pending
+				if err := k.SetVaultsV2UserPosition(ctx, address, position); err != nil {
+					return true, err
+				}
+			}
+
+			aggregatedDeposits, err = aggregatedDeposits.SafeAdd(pending)
+			if err != nil {
+				return true, err
+			}
+
+			return false, nil
+		}
+
+		numerator := new(big.Int).Mul(navBig, shares.BigInt())
+		quotient := new(big.Int)
+		remainder := new(big.Int)
+		quotient.QuoRem(numerator, totalSharesBig, remainder)
+
+		residual.Add(residual, remainder)
+		if residual.Cmp(totalSharesBig) >= 0 {
+			extra := new(big.Int).Div(new(big.Int).Set(residual), totalSharesBig)
+			if extra.Sign() > 0 {
+				quotient.Add(quotient, extra)
+				residual.Sub(residual, new(big.Int).Mul(totalSharesBig, extra))
+			}
+		}
+
+		value := sdkmath.NewIntFromBigInt(quotient)
+		accruedYield, err := value.SafeSub(shares)
+		if err != nil {
+			return true, err
+		}
+
+		position.AccruedYield = accruedYield
+		depositAmount := shares
+		if position.AmountPendingWithdrawal.IsPositive() {
+			if depositAmount, err = depositAmount.SafeAdd(position.AmountPendingWithdrawal); err != nil {
+				return true, err
+			}
+		}
+		position.DepositAmount = depositAmount
+		if err := k.SetVaultsV2UserPosition(ctx, address, position); err != nil {
+			return true, err
+		}
+
+		aggregatedYield, err = aggregatedYield.SafeAdd(accruedYield)
+		if err != nil {
+			return true, err
+		}
+
+		aggregatedDeposits, err = aggregatedDeposits.SafeAdd(depositAmount)
+		if err != nil {
+			return true, err
+		}
+
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	state.TotalAccruedYield = aggregatedYield
+	state.TotalDeposits = aggregatedDeposits
+	state.TotalNav = navInfo.CurrentNav
+	if !navInfo.LastUpdate.IsZero() {
+		state.LastNavUpdate = navInfo.LastUpdate
+	}
+
+	return k.SetVaultsV2VaultState(ctx, state)
+}
+
 // BeginBlocker is called at the beginning of each block.
 func (k *Keeper) BeginBlocker(ctx context.Context) error {
+	if err := k.updateVaultsV2Accounting(ctx); err != nil {
```

Comment:
Do we need to use a cached context in case of error to avoid partial state changes?